### DISCUSSION TITLE
Single cmd arg to reference models in existing A1111 checkout

### DISCUSSION THREAD

**Discussion by huchenlei** [lllyasviel/stable-diffusion-webui-forge#206] [Show and tell]
After https://github.com/lllyasviel/stable-diffusion-webui-forge/pull/203, you can use `--forge-ref-a1111-home` cmd arg to reference models in existing A1111 checkout.

Models include:
- Checkpoint
- VAE
- LoRA
- Embedding
- Hypernetwork
- ControlNet
- ControlNet preprocessor

Example webui-user.bat setup
```python
@echo off

set PYTHON=
set GIT=
set VENV_DIR=D:/stable-diffusion-webui/venv
set COMMANDLINE_ARGS=--forge-ref-a1111-home D:/stable-diffusion-webui

call webui.bat
```

**Comment by xhoxye** [comment]
I was wrong, the copied path did not modify the slash symbol
![QQ截图20240213184031](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/129571231/84e4a5d6-bbb8-4133-923a-cd7d043906c6)

**Comment by LiJT** [comment]
Please add ESRGAN and RealESRGAN or another upscaler model would be great! For now it doesnt read any of them

**Comment by dermesut** [comment]
this seems to work for me, except for the controlnet models.
when setting the "home", the path in which it looks is "{home}\models\ControlNet"
but the controlnet models of that "home"-installation are in "{home}\extensions\sd-webui-controlnet\models".
that's not a custom path that i put there myself. it's been there by default by that "home"-installation.
any chance to take that into account?

edit:
i believe, this is where that happens (launch_utils.py, line 521):
ModelRef(arg_name="--controlnet-dir", relative_path="models/ControlNet"),
        ModelRef(arg_name="--controlnet-preprocessor-models-dir", relative_path="extensions/sd-webui-controlnet/annotator/downloads"),

i don't understand, why the "correct" path (in terms of the extensions context) is there for the preprocessors, but not for the models.

btw: not joining parts of paths with os.path.join(a, b) but rather with things like target_path = a1111_home / ref.relative_path, gives me a hard time talking to your code programmatically.

but i'm not complaining. this forge-thingy is awesome. super-fast, let's me render huuuge images etc.
thank you very much for it!

**Reply by huchenlei** [reply]
If your ControlNet models location is not in `{home}\models\ControlNet`, you can override the default by specifying `--controlnet-dir` arg. 

The path you use was deprecated and we never remove support for it in sd-webui-controlnet. See https://github.com/Mikubill/sd-webui-controlnet/blob/9a5f28838a3670018fe00af48e4a3ee91582d279/scripts/global_state.py#L17-L18. That model path under extensions folder is marked with suffix `_old`.

**Reply by dermesut** [reply]
thank you.
not sure, if this is the place to ask, but what about lycoris?
i remember something along the lines of not needed anymore, treated like loras etc.. but i have some "old" one that are still working the way they used to before that (using "lyco:lyconame:value", that is) in the "classical" a1111.
with forge, they seem to have no effect at all.

**Comment by Renfield286** [comment]
What if your lora and checkpoints are on a different drive to the installation of A1111, is there a way to point forge at specific folders?

**Reply by seabean1999** [reply]
Some common command-line parameter:
 - `--ckpt-dir` - custom Stable-diffusion model dir
 - `--vae-dir` - custom VAE model dir
 - `--lora-dir` - custom LoRA model dir
 - `--controlnet-dir` - custom ControlNet model dir
 
 You could see more details in the file `modules/launch_utils.py` -> `configure_forge_reference_checkout -> refs`

**Comment by rm-000** [comment]
Hi everyone. I also need your help. I can't make Forge see my checkpoints. I'm trying to show Forge my Fooocus checkpoints, and I've tried many ways to write paths, but none worked.

Currently I have:

```
set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home D:/Fooocus_win64_2-1-831/Fooocus/

@REM Uncomment following code to reference an existing A1111 checkout.
FOOOCUS_HOME=D:/Fooocus_win64_2-1-831/Fooocus/

set VENV_DIR=%FOOOCUS_HOME%/venv
set COMMANDLINE_ARGS=%COMMANDLINE_ARGS% ^
--ckpt-dir %FOOOCUS_HOME%/models/checkpoints ^
--hypernetwork-dir %FOOOCUS_HOME%/models/hypernetworks ^
--embeddings-dir %FOOOCUS_HOME%/embeddings ^
--lora-dir %FOOOCUS_HOME%/models/loras
```

I tried / and \ slashes in COMMANDLINE_ARGS and FOOOCUS_HOME. I also wrote 'models' like this:
D:/Fooocus_win64_2-1-831/Fooocus/models
It didn't work either.

What I get in cmd is:
```
Path D:\Fooocus_win64_2-1-831\Fooocus\models\Stable-diffusion does not exist. Skip setting --ckpt-dir
Path D:\Fooocus_win64_2-1-831\Fooocus\embeddings does not exist. Skip setting --embeddings-dir
Path D:\Fooocus_win64_2-1-831\Fooocus\models\lora does not exist. Skip setting --lora-dir
Path D:\Fooocus_win64_2-1-831\Fooocus\extensions\sd-webui-controlnet\annotator\downloads does not exist. Skip setting --controlnet-preprocessor-models-dir
```

and WebUI doesn't see my checkpoints. Can anyone please say how to fix this?

**Reply by SvenErik1968** [reply]
Try changing your line:
`FOOOCUS_HOME=D:/Fooocus_win64_2-1-831/Fooocus/`
to:
`set A1111_HOME=D:\Fooocus_win64_2-1-831\Fooocus\`

You had not used the **`SET`** command when you defined the HOME-folder, I would also guess the Forge-code will look for the `A1111_HOME` variable and not `FOOOCUS_HOME`.

Another way to fix this, especially if you are planning to use your checkpoints and other kinds of models with several programs, is to create a separate folder-system outside any programs,
```
D:\Models\
D:\Models\Checkpoints\SDXL 1.0\
D:\Models\Checkpoints\Pony\
D:\Models\Checkpoints\SD 1.5\
D:\Models\Lora\SDXL 1.0\
D:\Models\Lora\Pony\
D:\Models\Lora\SD 1.5\
```
etc...

and then create symlinks to those folders as needed within the various programs you are using. Decent guide for this [here on CivitAI](https://civitai.com/articles/4141/tutorial-how-to-use-your-stable-diffusion-models-without-copying-them).

**Reply by rm-000** [reply]
Thanks for your help, and link too.

Yes, I am actually trying to make a common folder for different programs, WebUI included.
While I'm trying to get this thing working using the info in your link, could you please explain, why is this not working?
I did as you wrote, but it doesn't work.

This is what I have now:

```
@echo off

set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home D:/Soft/AI/Fooocus 2.5.5/Fooocus/

@REM Uncomment following code to reference an existing A1111 checkout.
set A1111_HOME=D:\Fooocus 2.5.5\Fooocus\

set VENV_DIR=%A1111_HOME%/venv
set COMMANDLINE_ARGS=%COMMANDLINE_ARGS% ^
--ckpt-dir %A1111_HOME%/models/checkpoints ^
--hypernetwork-dir %A1111_HOME%/models/hypernetworks ^
--embeddings-dir %A1111_HOME%/embeddings ^
--lora-dir %A1111_HOME%/models/loras

call webui.bat

```

For some reason it goes to a wrong direction (doesn't look for Fooocus 2.5.5) and doesn't start:

```
Path D:\Soft\AI\Fooocus\models\Stable-diffusion does not exist. Skip setting --ckpt-dir
Path D:\Soft\AI\Fooocus\models\VAE does not exist. Skip setting --vae-dir
Path D:\Soft\AI\Fooocus\models\hypernetworks does not exist. Skip setting --hypernetwork-dir
Path D:\Soft\AI\Fooocus\embeddings does not exist. Skip setting --embeddings-dir
Path D:\Soft\AI\Fooocus\models\lora does not exist. Skip setting --lora-dir
Path D:\Soft\AI\Fooocus\models\ControlNet does not exist. Skip setting --controlnet-dir
Path D:\Soft\AI\Fooocus\extensions\sd-webui-controlnet\annotator\downloads does not exist. Skip setting --controlnet-preprocessor-models-dir
Launching Web UI with arguments: --theme dark --forge-ref-a1111-home D:/Soft/AI/Fooocus 2.5.5/Fooocus/ --ckpt-dir D:Fooocus 2.5.5Fooocus/models/checkpoints --hypernetwork-dir D:Fooocus 2.5.5Fooocus/models/hypernetworks --embeddings-dir D:Fooocus 2.5.5Fooocus/embeddings --lora-dir D:Fooocus 2.5.5Fooocus/models/loras
...
launch.py: error: unrecognized arguments: 2.5.5/Fooocus/ 2.5.5Fooocus/models/checkpoints 2.5.5Fooocus/models/hypernetworks 2.5.5Fooocus/embeddings 2.5.5Fooocus/models/loras
Press any key to continue . . .
```

**Comment by dermesut** [comment]
try using quotation marks with your forge-home path

**Reply by rm-000** [reply]
Thanks for the suggestion. I've tried these:

set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home "D:/Soft/AI/Fooocus 2.5.5/Fooocus/"
set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home 'D:/Soft/AI/Fooocus 2.5.5/Fooocus/'
set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home 'D:\Soft\AI\Fooocus 2.5.5\Fooocus\'
set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home "D://Soft//AI//Fooocus 2.5.5//Fooocus//"
set COMMANDLINE_ARGS=--theme dark --forge-ref-a1111-home D://Soft//AI//Fooocus 2.5.5//Fooocus//

None work

**Comment by rm-000** [comment]
Finally! I got it working! Cudos to the author of this article:
https://www.digitalcreativeai.net/en/post/how-use-flux-1-dev-webui-forge-how-use-gguf-and-lora

The working solution for me to show path to an outside folder was to change just the Commandline_Args like this:
`set COMMANDLINE_ARGS=--theme darkset --ckpt-dir "D:\Soft\AI\test\models"`

While playing with the settings, I tried different things and noticed that using, for example, Russian words:
`set COMMANDLINE_ARGS=--theme darkset --ckpt-dir "D:\Soft\AI\тест\models"`

would return a strange log:
```
_PYTHON: can't open file 'D:\\Soft\\AI\\Webui Forge\\webui\\=': [Errno 2] No such file or directory
git: '=' is not a git command. See 'git --help'.

The most similar commands are
        am
        gc
        mv
        rm
'VENV_DIR' is not recognized as an internal or external command,
operable program or batch file.
'COMMANDLINE_ARGS' is not recognized as an internal or external command,
operable program or batch file._
Python 3.10.6 (...)
```

Using a space in the address line like this:
`set COMMANDLINE_ARGS=--ckpt-dir "D:\Soft\AI\**test test**\models"`
_wouldn't_ crash WebUI.

To use all checkpoints, loras, embeddings, vae write this:
set COMMANDLINE_ARGS=--ckpt-dir "C:\your-folder\Stable-diffusion" --lora-dir "C:\your-folder\Lora" --embeddings-dir "C:\your-folder\embeddings" --vae-dir "C:\your-folder\VAE"

Thank you for everyone's involvement!
